service/controller/v2/resource/vnetpeering/current.go

```diff
@@ -2,40 +2,86 @@ package vnetpeering
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2017-09-01/network"
-	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
 	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
 
 	"github.com/giantswarm/azure-operator/client"
 	"github.com/giantswarm/azure-operator/service/controller/v2/key"
 )
 
-// GetCurrentState retrieve the current host cluster virtual network peering resource from azure.
-func (r Resource) GetCurrentState(ctx context.Context, azureConfig interface{}) (interface{}, error) {
-	a, err := key.ToCustomObject(azureConfig)
+// GetCurrentState retrieve the current host cluster virtual network peering
+// resource from azure.
+func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
+	customObject, err := key.ToCustomObject(obj)
 	if err != nil {
 		return network.VirtualNetworkPeering{}, microerror.Mask(err)
 	}
 
-	vnetPeering, err := r.getCurrentState(ctx, a)
-	if err != nil {
-		return network.VirtualNetworkPeering{}, microerror.Mask(err)
-	}
+	// In order to make vnet peering work we need a virtual network which we can
+	// use to peer. In case there is no virtual network yet we cancel the resource
+	// and try again on the next resync period. This is a classical scenario on
+	// guest cluster creation. If we would not check for the virtual network
+	// existence the client calls of CreateOrUpdate would fail with not found
+	// errors.
+	{
+		c, err := r.getVirtualNetworksClient()
+		if err != nil {
+			return network.VirtualNetworkPeering{}, microerror.Mask(err)
+		}
 
-	return vnetPeering, nil
-}
+		g := key.ResourceGroupName(customObject)
+		n := key.VnetName(customObject)
+		e := ""
+		v, err := c.Get(ctx, g, n, e)
+		if IsVirtualNetworkNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find the virtual network in the Azure API")
+			resourcecanceledcontext.SetCanceled(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
 
-func (r Resource) getCurrentState(ctx context.Context, azureConfig providerv1alpha1.AzureConfig) (network.VirtualNetworkPeering, error) {
-	vnetPeeringClient, err := r.getVnetPeeringClient()
-	if err != nil {
-		return network.VirtualNetworkPeering{}, microerror.Mask(err)
+			return network.VirtualNetworkPeering{}, nil
+		} else if err != nil {
+			return network.VirtualNetworkPeering{}, microerror.Mask(err)
+		} else {
```

Comment:
This `else` is not needed, there are `return` in all other branches.

Reply:
I keep this for consistency.